File: Trees/Java/src/HeapDemo.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.PriorityQueue;

public class HeapDemo {
    private static final Path DEFAULT_QUERY_FILE = Paths.get("..", "test_case", "heap_queries.txt");

    private static class MinHeap {
        private PriorityQueue<Integer> heap = new PriorityQueue<>();

        void push(int value) {
            heap.offer(value);
        }

        Integer pop() {
            return heap.poll();
        }

        Integer peek() {
            return heap.peek();
        }

        void heapify(List<Integer> values) {
            heap = new PriorityQueue<>(values);
        }

        int size() {
            return heap.size();
        }

        List<Integer> values() {
            return new ArrayList<>(heap);
        }
    }

    private static String formatValues(List<Integer> values) {
        if (values.isEmpty()) {
            return "(empty)";
        }
        StringBuilder builder = new StringBuilder();
        for (int i = 0; i < values.size(); i++) {
            if (i > 0) {
                builder.append(' ');
            }
            builder.append(values.get(i));
        }
        return builder.toString();
    }

    public static void main(String[] args) throws IOException {
        Path queryFile = args.length > 0 ? Paths.get(args[0]) : DEFAULT_QUERY_FILE;
        if (!Files.isRegularFile(queryFile)) {
            System.err.println("Query file not found: " + queryFile);
            return;
        }

        MinHeap heap = new MinHeap();
        System.out.println("Heap / Priority Queue Demo (Java)");
        System.out.println("=================================");

        for (String rawLine : Files.readAllLines(queryFile)) {
            String line = rawLine.trim();
            if (line.isEmpty() || line.startsWith("#")) {
                continue;
            }

            String[] parts = line.split("\\s+");
            String command = parts[0].toLowerCase();

            switch (command) {
                case "push" -> {
                    int value = Integer.parseInt(parts[1]);
                    heap.push(value);
                    System.out.println("push " + value + " -> OK");
                }
                case "pop" -> {
                    Integer value = heap.pop();
                    System.out.println("pop -> " + (value == null ? "(empty)" : value));
                }
                case "peek" -> {
                    Integer value = heap.peek();
                    System.out.println("peek -> " + (value == null ? "(empty)" : value));
                }
                case "size" -> System.out.println("size -> " + heap.size());
                case "print" -> System.out.println("heap -> " + formatValues(heap.values()));
                case "heapify" -> {
                    List<Integer> values = new ArrayList<>();
                    for (int i = 1; i < parts.length; i++) {
                        values.add(Integer.parseInt(parts[i]));
                    }
                    heap.heapify(values);
                    System.out.println("heapify -> " + formatValues(heap.values()));
                }
                default -> System.out.println("unknown command: " + line);
            }
        }
    }
}

File: Trees/Python/heap_demo.py
from __future__ import annotations

import heapq
import sys
from pathlib import Path

TEST_CASE_FILE = Path("..") / "test_case" / "heap_queries.txt"


class MinHeap:
    def __init__(self) -> None:
        self.data: list[int] = []

    def push(self, value: int) -> None:
        heapq.heappush(self.data, value)

    def pop(self) -> int | None:
        if not self.data:
            return None
        return heapq.heappop(self.data)

    def peek(self) -> int | None:
        if not self.data:
            return None
        return self.data[0]

    def heapify(self, values: list[int]) -> None:
        self.data = values[:]
        heapq.heapify(self.data)

    def size(self) -> int:
        return len(self.data)

    def values(self) -> list[int]:
        return self.data[:]


def format_values(values: list[int]) -> str:
    return " ".join(str(value) for value in values) if values else "(empty)"


def run_queries(file_path: Path) -> int:
    if not file_path.is_file():
        print(f"Query file not found: {file_path}", file=sys.stderr)
        return 1

    heap = MinHeap()
    print("Heap / Priority Queue Demo (Python)")
    print("===================================")

    for raw_line in file_path.read_text().splitlines():
        line = raw_line.strip()
        if not line or line.startswith("#"):
            continue

        parts = line.split()
        command = parts[0].lower()

        if command == "push":
            value = int(parts[1])
            heap.push(value)
            print(f"push {value} -> OK")
        elif command == "pop":
            value = heap.pop()
            print(f"pop -> {value if value is not None else '(empty)'}")
        elif command == "peek":
            value = heap.peek()
            print(f"peek -> {value if value is not None else '(empty)'}")
        elif command == "size":
            print(f"size -> {heap.size()}")
        elif command == "print":
            print(f"heap -> {format_values(heap.values())}")
        elif command == "heapify":
            values = [int(token) for token in parts[1:]]
            heap.heapify(values)
            print(f"heapify -> {format_values(heap.values())}")
        else:
            print(f"unknown command: {line}")

    return 0


def main() -> int:
    query_file = TEST_CASE_FILE
    if len(sys.argv) > 1:
        query_file = Path(sys.argv[1])
    return run_queries(query_file)


if __name__ == "__main__":
    raise SystemExit(main())
